### PR TITLE
improved repr

### LINKED ISSUE #4: Improve __repr__
- leave out properties that are set to the default
- summarize containers rather than recursively include their contents.

possible example `repr(OME)`:
```
OME(uuid='urn:uuid:afa03d8b-30a8-4c9b-b016-9992c3d73fef', image=<1 image>)
```
e.g `Pixels`
```
   Pixels(
     dimension_order=DimensionOrder.XYZCT, id='Pixels:0',
     size_c=8, size_t=1, size_x=7218, size_y=4078, size_z=1,
     type=PixelType.UINT16,    physical_size_x=0.32499998807907104,
     physical_size_x_unit=UnitsLength.MICROMETER,
     physical_size_y=0.32499998807907104,
     physical_size_y_unit=UnitsLength.MICROMETER,
     channel=<8 Channels>, tiff_data=<1 TiffData>, plane=<8 Planes>,
   )
```

### PR DESCRIPTION
**Pull request by tlambert03**
closes #4
This is a start on improved reprs, as per #4 (leave out properties that are set to the default, summarize containers rather than recursively include their contents) plenty to discuss here...
here's an `Image` repr as it stands:

```
Image:0
   pixels=Pixels:0:0
      dimension_order='XYCZT'
      size_c=2
      size_t=2
      size_x=6
      size_y=4
      size_z=2
      type='uint8'
      bin_data=<8 Bin_Data>
      channels=<2 Channels>
      physical_size_x=10000.0
      physical_size_y=10000.0
      planes=<8 Planes>
   acquisition_date=2010-11-10 14:42:38
   annotation_ref=<4 Annotation_Ref>
   instrument_ref=Instrument:SpimSampleMicroscope1

   name=Spim Sample Tile 1 Angle 1
   objective_settings=Objective:1
      correction_collar=6.0
      medium='Oil'
   stage_label=StageLabel
      name=(1,1) of 1x2
      x=1.0
      y=1.0
```

<details>

<summary>previously...</summary>

:joy:

```
Image(pixels=Pixels(dimension_order=<DimensionOrder.XYCZT: 'XYCZT'>,
size_c=2, size_t=2, size_x=6, size_y=4, size_z=2, type=<PixelType.UINT8:
'uint8'>, big_endian=None,
bin_data=[BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False,
length=32, compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>),
BinData(value='/wCrzur//wB5oMPi/wBIbJO3AP8ePGCF', big_endian=False, length=32,
compression=<Compression.NONE: 'none'>)],
channels=[Channel(acquisition_mode=None, annotation_ref=[], color=-1,
contrast_method=None, detector_settings=None, emission_wavelength=None,
emission_wavelength_unit=<UnitsLength.NANOMETER: 'nm'>,
excitation_wavelength=None, excitation_wavelength_unit=<UnitsLength.NANOMETER:
'nm'>, filter_set_ref=None, fluor='Autofluorescence', id='Channel:0.0',
illumination_type=None, light_path=None, light_source_settings=None, name=None,
nd_filter=None, pinhole_size=None, pinhole_size_unit=<UnitsLength.MICROMETER:
'µm'>, pockel_cell_setting=None, samples_per_pixel=None),
Channel(acquisition_mode=None, annotation_ref=[], color=16711935,
contrast_method=None, detector_settings=None, emission_wavelength=None,
emission_wavelength_unit=<UnitsLength.NANOMETER: 'nm'>,
excitation_wavelength=None, excitation_wavelength_unit=<UnitsLength.NANOMETER:
'nm'>, filter_set_ref=None, fluor='Green-OME', id='Channel:0.1',
illumination_type=None, light_path=None, light_source_settings=None, name=None,
nd_filter=None, pinhole_size=None, pinhole_size_unit=<UnitsLength.MICROMETER:
'µm'>, pockel_cell_setting=None, samples_per_pixel=None)], id='Pixels:0:0',
interleaved=None, metadata_only=False, physical_size_x=10000.0,
physical_size_x_unit=<UnitsLength.MICROMETER: 'µm'>, physical_size_y=10000.0,
physical_size_y_unit=<UnitsLength.MICROMETER: 'µm'>, physical_size_z=None,
physical_size_z_unit=<UnitsLength.MICROMETER: 'µm'>, planes=[Plane(the_c=0,
the_t=0, the_z=0, annotation_ref=[], delta_t=None,
delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=1, the_t=0, the_z=0, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=0, the_t=0, the_z=1, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=1, the_t=0, the_z=1, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=0, the_t=1, the_z=0, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=1, the_t=1, the_z=0, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=0, the_t=1, the_z=1, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>), Plane(the_c=1, the_t=1, the_z=1, annotation_ref=[],
delta_t=None, delta_t_unit=<UnitsTime.SECOND: 's'>, exposure_time=500.0,
exposure_time_unit=<UnitsTime.SECOND: 's'>, hash_sha1=None, position_x=None,
position_x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>,
position_y=None, position_y_unit=<UnitsLength.REFERENCEFRAME: 'reference
frame'>, position_z=None, position_z_unit=<UnitsLength.REFERENCEFRAME:
'reference frame'>)], significant_bits=None, tiff_data_blocks=[],
time_increment=None, time_increment_unit=<UnitsTime.SECOND: 's'>),
acquisition_date=datetime.datetime(2010, 11, 10, 14, 42, 38),
annotation_ref=[AnnotationRef(id='Annotation:ExtraStageLabel:1:0'),
AnnotationRef(id='Annotation:SpimSet:1'),
AnnotationRef(id='Annotation:ObjectiveAdditions:1'),
AnnotationRef(id='Annotation:SystemSpecific:1')], description=None,
experiment_ref=None, experimenter_group_ref=None, experimenter_ref=None,
id='Image:0', imaging_environment=None,
instrument_ref=InstrumentRef(id='Instrument:SpimSampleMicroscope1'),
microbeam_manipulation_ref=[], name='Spim Sample Tile 1 Angle 1',
objective_settings=ObjectiveSettings(correction_collar=6.0, id='Objective:1',
medium=<Medium.OIL: 'Oil'>, refractive_index=None), roi_ref=[],
stage_label=StageLabel(name='(1,1) of 1x2', x=1.0,
x_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>, y=1.0,
y_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>, z=None,
z_unit=<UnitsLength.REFERENCEFRAME: 'reference frame'>))
```

</details>

### CODE REVIEW

**Comment by jmuhlich** [comment]
I think this is really close to creating reprs that are basically executable Python code (except for the containers)! This kind of repr would go a long way in guiding new users down the right path for creating their own objects without even needing to consult the docs. This near-executable output might look like this for your example:

```python
Image(
   id="Image:0",
   pixels=Pixels(
      id="Pixels:0:0",
      dimension_order="XYCZT",
      size_c=2,
      size_t=2,
      size_x=6,
      size_y=4,
      size_z=2,
      type="uint8",
      bin_data=<8 Bin_Data>,
      channels=<2 Channels>,
      physical_size_x=10000.0,
      physical_size_y=10000.0,
      planes=<8 Planes>,
   ),
   acquisition_date=datetime.fromisoformat("2010-11-10T14:42:38"),
   annotation_ref=<4 Annotation_Ref>,
   instrument_ref=InstrumentRef(id="Instrument:SpimSampleMicroscope1"),
   name="Spim Sample Tile 1 Angle 1",
   objective_settings=ObjectiveSettings(
      id="Objective:1",
      correction_collar=6.0,
      medium="Oil",
   ),
   stage_label=StageLabel(
      name="(1,1) of 1x2",
      x=1.0,
      y=1.0,
   ),
)
```

**Comment by tlambert03** [comment]
yep, love that... will work towards it.

**Comment by tlambert03** [comment]
updated... some examples:

```python
Image(
   id='Image:0',
   name='Spim Sample Tile 1 Angle 1',
   pixels=Pixels(
      id='Pixels:0:0',
      dimension_order='XYCZT',
      size_c=2,
      size_t=2,
      size_x=6,
      size_y=4,
      size_z=2,
      type='uint8',
      bin_data=[<8 Bin_Data>],
      channels=[<2 Channels>],
      physical_size_x=10000.0,
      physical_size_y=10000.0,
      planes=[<8 Planes>]
   ),
   acquisition_date=datetime.fromisoformat('2010-11-10T14:42:38'),
   annotation_ref=[<4 Annotation_Ref>],
   instrument_ref=InstrumentRef(
      id='Instrument:SpimSampleMicroscope1'
   ),
   objective_settings=ObjectiveSettings(
      id='Objective:1',
      correction_collar=6.0,
      medium='Oil'
   ),
   stage_label=StageLabel(
      name='(1,1) of 1x2',
      x=1.0,
      y=1.0
   )
)

OME(
   images=[<4 Images>],
   instruments=[<1 Instruments>],
   structured_annotations=[<7 Structured_Annotations>]
)

Instrument(
   id='Instrument:SpimSampleMicroscope1',
   detectors=[<1 Detectors>],
   light_source_group=[<2 Light_Source_Group>],
   microscope=Microscope(
      manufacturer='OME-Sample',
      model='Spim-Sample',
      serial_number='SPIM-1',
      type='Other'
   ),
   objectives=[<3 Objectives>]
)

Laser(
   id='LightSource:1',
   manufacturer='OME-Sample',
   model='Laser Mk1',
   serial_number='LASER-1',
   laser_medium='GaAs',
   type='Semiconductor',
   wavelength=448.0
)
```

File: src/ome_types/dataclasses.py [hunk 1]
```diff
@@ -63,11 +78,52 @@ def new_post_init(self: Any, *args: Any) -> None:
...
+                    rep = f"[<{len(current)} {f.name.title()}>]"
+                elif isinstance(current, Enum):
+                    rep = repr(current.value)
+                elif isinstance(current, datetime):
+                    rep = f"datetime.fromisoformat({current.isoformat()!r})"
+                else:
+                    rep = repr(current)
+                lines.append(f"{f.name}={rep},")
+        if lines:
+            lines[-1] = lines[-1].rstrip(",")
```
Comment (jmuhlich) [inline src/ome_types/dataclasses.py:112]: Trailing commas are OK in the Python grammar and I think the Black style requires them anyway. I would support removing this line to leave the trailing comma in.

Comment (tlambert03) [reply src/ome_types/dataclasses.py:112]: Ok, I think I’ll make it `if len(lines) > 1` then... since a comma on a one-liner bugs me :)

Comment (jmuhlich) [reply src/ome_types/dataclasses.py:112]: OK, easier to add more lines after it (e.g. after copy-pasting to your own code) if it already has the comma though!

File: src/ome_types/dataclasses.py [hunk 2]
```diff
@@ -63,11 +78,52 @@ def new_post_init(self: Any, *args: Any) -> None:
...
+            body = "\n" + indent("\n".join(lines), "   ") + "\n"
```
Comment (jmuhlich) [inline src/ome_types/dataclasses.py:113]: Oops this should be a 4-space line. Consider `" " * 4` for more clarity.

File: src/ome_types/dataclasses.py (hunk 1 above)
Comment (tlambert03) [reply src/ome_types/dataclasses.py:112]: whoops... sorry, I didn't mean `len(lines)>1`, I mean `len(lines)==1`.  I agree with you about the trailing comma... just don't like single lines to be: `instrument_ref=InstrumentRef(id='Instrument:1',)`

**Pull request merged**